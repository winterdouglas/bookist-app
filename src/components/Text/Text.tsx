import { ReactNode } from "react";
import {
  type StyleProp,
  Text as RNText,
  type TextProps as RNTextProps,
  type TextStyle,
} from "react-native";
import { useAppTheme } from "@/hooks/useAppTheme";

type Sizes = keyof typeof $sizeStyles;
type Presets = keyof typeof $presets;

export type TextProps = RNTextProps & {
  text?: string;
  preset?: Presets;
  size?: Sizes;
  children?: ReactNode;
};

export const Text = ({
  preset = "default",
  size,
  text,
  children,
  style: $styleOverride,
  ...rest
}: TextProps) => {
  const { colors } = useAppTheme();
  const content = text || children;

  const $styles = [
    $presets[preset],
    size && $sizeStyles[size],
    { color: colors.text },
    $styleOverride,
  ];

  return (
    <RNText selectable {...rest} style={$styles}>
      {content}
    </RNText>
  );
};

const $sizeStyles = {
  lg: { fontSize: 28 } as TextStyle,
  md: { fontSize: 20 } as TextStyle,
  sm: { fontSize: 16 } as TextStyle,
  xs: { fontSize: 12 } as TextStyle,
};

const $baseStyle: StyleProp<TextStyle> = [
  $sizeStyles.sm,
  { fontWeight: "500" },
];

const $presets = {
  default: $baseStyle,
  heading: [
    $baseStyle,
    $sizeStyles.lg,
    { fontWeight: 900 },
  ] as StyleProp<TextStyle>,
  subheading: [
    $baseStyle,
    $sizeStyles.md,
    { fontWeight: 900 },
  ] as StyleProp<TextStyle>,
  list: [$baseStyle, { fontWeight: 600 }] as StyleProp<TextStyle>,
  subtitle: [
    $baseStyle,
    $sizeStyles.xs,
    { fontWeight: 700 },
  ] as StyleProp<TextStyle>,
};
