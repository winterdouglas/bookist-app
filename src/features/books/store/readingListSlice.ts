import { RootState } from "@/store";
import { withPayloadType } from "@/utils/withPayloadType";
import { createAction, createSlice } from "@reduxjs/toolkit";

export const toggleReadingBook = createAction(
  "readingList/bookToggled",
  withPayloadType<string>(),
);

export const readingListSlice = createSlice({
  name: "readingList",
  initialState: {
    readingBooksByBookId: {} as Record<string, boolean>,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(toggleReadingBook, (state, action) => {
      const bookId = action.payload;
      state.readingBooksByBookId[bookId] = !state.readingBooksByBookId[bookId];
    });
  },
});

export const selectIsReadingBook = (state: RootState, bookId: string) =>
  state.readingList.readingBooksByBookId[bookId] ?? false;
