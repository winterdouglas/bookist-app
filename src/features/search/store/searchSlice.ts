import {
  createAsyncThunk,
  createSelector,
  createSlice,
} from "@reduxjs/toolkit";
import type { AppAsyncThunkConfig, RequestStatus, RootState } from "@/store";
import { SearchResult } from "@/features/search/models";
import { SearchApi } from "@/features/search/services";

type StatusBySearchTerm = Record<string, RequestStatus | undefined>;

type DataBySearchTerm = Record<
  string,
  | {
      nextPage: number;
      allFetched: boolean;
      lastRequestedTime?: Record<number, number>;
      pages: Record<number, SearchResult[]>;
    }
  | undefined
>;

type ItemPageBySearchTerm = Record<string, Record<string, number>>;

type SearchBooksThunkResult = {
  results: SearchResult[];
  page: number;
};

export const searchBooks = createAsyncThunk<
  SearchBooksThunkResult,
  string,
  AppAsyncThunkConfig
>(
  "books/search",
  async (searchTerm, { getState, signal }) => {
    const state = getState();
    const page = selectPageBySearchTerm(state, searchTerm);
    const data = selectDataBySearchTerm(state, searchTerm);

    // TODO: Cache by date
    const cachedData = data?.pages[page];
    if (cachedData) {
      // Use cached data
      return { results: cachedData, page };
    }

    const result = await SearchApi.searchBooks(searchTerm, page, signal);
    return { results: result.docs, page };
  },
  {
    condition: (searchTerm, { getState }) => {
      const state = getState();
      const status = selectStatusBySearchTerm(state, searchTerm);
      const allFetched = selectAllFetchedBySearchTerm(state, searchTerm);

      return searchTerm.length >= 3 && status !== "pending" && !allFetched;
    },
  },
);

export const searchSlice = createSlice({
  name: "search",
  initialState: {
    dataBySearchTerm: {} as DataBySearchTerm,
    statusBySearchTerm: {} as StatusBySearchTerm,
    itemPageBySearchTerm: {} as ItemPageBySearchTerm,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(searchBooks.pending, (state, action) => {
      const searchTerm = action.meta.arg;
      state.statusBySearchTerm[searchTerm] = action.meta.requestStatus;
    });
    builder.addCase(searchBooks.fulfilled, (state, action) => {
      const searchTerm = action.meta.arg;
      const { results, page } = action.payload;
      const nextPage = results.length > 0 ? page + 1 : page;

      // TODO: Check
      state.dataBySearchTerm[searchTerm] = {
        ...(state.dataBySearchTerm[searchTerm] ?? {}),
        nextPage: nextPage,
        allFetched: results.length === 0,
        lastRequestedTime: {
          ...(state.dataBySearchTerm[searchTerm]?.lastRequestedTime ?? {}),
          [page]: Date.now(),
        },
        pages: {
          ...(state.dataBySearchTerm[searchTerm]?.pages ?? {}),
          [page]: results,
        },
      };
      state.statusBySearchTerm[searchTerm] = action.meta.requestStatus;
      state.itemPageBySearchTerm[searchTerm] = results.reduce(
        (acc, current) => {
          acc = { ...acc, [current.key]: page };
          return acc;
        },
        state.itemPageBySearchTerm[searchTerm] as Record<string, number>,
      );
    });
    builder.addCase(searchBooks.rejected, (state, action) => {
      const searchTerm = action.meta.arg;
      state.statusBySearchTerm[searchTerm] = action.meta.requestStatus;
    });
  },
});

export const selectPageBySearchTerm = (state: RootState, searchTerm: string) =>
  state.search.dataBySearchTerm[searchTerm]?.nextPage ?? 1;

export const selectAllFetchedBySearchTerm = (
  state: RootState,
  searchTerm: string,
) => state.search.dataBySearchTerm[searchTerm]?.allFetched;

export const selectDataBySearchTerm = (state: RootState, searchTerm: string) =>
  state.search.dataBySearchTerm[searchTerm];

export const selectStatusBySearchTerm = (
  state: RootState,
  searchTerm: string,
) => state.search.statusBySearchTerm[searchTerm];

export const selectAllPagesBySearchTerm = createSelector(
  selectDataBySearchTerm,
  selectPageBySearchTerm,
  (data, page) => {
    if (!data || !data.pages || !page) {
      return [];
    }

    return Object.entries(data.pages)
      .map(([key, value]) => [parseInt(key, 10), value] as const)
      .sort(([k1], [k2]) => k1 - k2)
      .reduce((acc, [currKey, currValue]) => {
        if (currKey <= page) {
          acc = [...acc, ...currValue];
        }
        return acc;
      }, [] as SearchResult[]);
  },
);

export const selectSearchResultPage = (
  state: RootState,
  searchTerm: string,
  itemId: string,
) => state.search.itemPageBySearchTerm[searchTerm]?.[itemId];

export const selectSearchResult = (
  state: RootState,
  searchTerm: string,
  page: number,
  itemId: string,
) =>
  // TODO: Make the stored data keyed
  state.search.dataBySearchTerm[searchTerm]?.pages[page]?.find(
    (x) => x.key === itemId,
  );
