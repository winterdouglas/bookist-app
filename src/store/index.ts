export * from "./setupStore";
export * from "./reactNativeListeners";
