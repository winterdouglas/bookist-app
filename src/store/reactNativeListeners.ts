import { AppState, NativeEventSubscription } from "react-native";
import NetInfo, { NetInfoSubscription } from "@react-native-community/netinfo";
import { ThunkDispatch } from "@reduxjs/toolkit";
import {
  onFocus as onFocusType,
  onFocusLost as onFocusLostType,
  onOnline as onOnlineType,
  onOffline as onOfflineType,
} from "@reduxjs/toolkit/dist/query/core/setupListeners";

type CustomHandlerActions = {
  onFocus: typeof onFocusType;
  onFocusLost: typeof onFocusLostType;
  onOnline: typeof onOnlineType;
  onOffline: typeof onOfflineType;
};

let initialized = false;

export const reactNativeListeners = (
  dispatch: ThunkDispatch<any, any, any>,
  { onFocus, onFocusLost, onOnline, onOffline }: CustomHandlerActions,
) => {
  let unsubscribeOnChange: NativeEventSubscription | undefined;
  let unsubscribeOnNetworkStatusChange: NetInfoSubscription | undefined;

  if (!initialized) {
    // Handle focus events
    unsubscribeOnChange = AppState.addEventListener("change", (state) => {
      if (state === "active") {
        dispatch(onFocus());
      } else if (state === "background") {
        dispatch(onFocusLost());
      }
    });

    // Handle connection events
    unsubscribeOnNetworkStatusChange = NetInfo.addEventListener((state) => {
      if (state.isConnected) {
        dispatch(onOnline());
      } else {
        dispatch(onOffline());
      }
    });
    initialized = true;
  }

  const unsubscribe = () => {
    unsubscribeOnChange?.remove();
    unsubscribeOnNetworkStatusChange?.();
    initialized = false;
  };
  return unsubscribe;
};
