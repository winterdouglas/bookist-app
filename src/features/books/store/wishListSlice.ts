import { RootState } from "@/store";
import { withPayloadType } from "@/utils/withPayloadType";
import { createAction, createSlice } from "@reduxjs/toolkit";

export const toggleWishedBook = createAction(
  "wishList/bookToggled",
  withPayloadType<string>(),
);

export const wishListSlice = createSlice({
  name: "wishList",
  initialState: {
    wishedBooksByBookId: {} as Record<string, boolean>,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(toggleWishedBook, (state, action) => {
      const bookId = action.payload;
      state.wishedBooksByBookId[bookId] = !state.wishedBooksByBookId[bookId];
    });
  },
});

export const selectIsWishedBook = (state: RootState, bookId: string) =>
  state.wishList.wishedBooksByBookId[bookId] ?? false;
