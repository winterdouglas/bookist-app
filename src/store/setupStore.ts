import {
  combineReducers,
  configureStore,
  createAsyncThunk,
} from "@reduxjs/toolkit";
import { searchSlice } from "@/features/search/store/searchSlice";

const reducer = combineReducers({
  search: searchSlice.reducer,
});

export const setupStore = () => {
  return configureStore({
    reducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
  });
};

export type RootState = ReturnType<typeof reducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore["dispatch"];

export type AppAsyncThunkConfig<TExtra = unknown, TRejectedValue = unknown> = {
  state: RootState;
  dispatch: AppDispatch;
  extra?: TExtra;
  rejectValue?: TRejectedValue;
  serializedErrorType?: unknown;
  pendingMeta?: unknown;
  fulfilledMeta?: unknown;
  rejectedMeta?: unknown;
};

export type RequestStatus = "pending" | "fulfilled" | "rejected";
