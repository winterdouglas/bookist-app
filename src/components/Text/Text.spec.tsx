import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { setupStore } from "@/store";
import { Text } from "./Text";

it("renders correctly", () => {
  const { store } = setupStore();
  const component = renderer.create(
    <Provider store={store}>
      <Text text="just some fake text!" />
    </Provider>,
  );

  expect(component.toJSON()).toMatchSnapshot();
});
