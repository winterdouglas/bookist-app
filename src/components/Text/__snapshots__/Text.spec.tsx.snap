// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  selectable={true}
  style={
    [
      [
        {
          "fontSize": 16,
        },
        {
          "fontWeight": "500",
        },
      ],
      undefined,
      {
        "color": "rgb(28, 28, 30)",
      },
      undefined,
    ]
  }
>
  just some fake text!
</Text>
`;
